### PR TITLE
[WIP] Support RGF (Fully corrective update)

### PR DESCRIPTION
**Pull request by fukatani**
See also #315 
Especially, it is valuable to introduce fully corrective update to LightGBM, I think.

Debugging and performance verification are in progress.
Maybe, I will move forward slowly.

Any opinion and question is welcome.

### CODE REVIEW

**Comment by Laurae2** [comment]
@fukatani Can you sign the CLA?

File: CMakeLists.txt [hunk 1]
```diff
@@ -97,7 +97,7 @@ if(USE_HDFS)
 endif(USE_HDFS)
 
 if(UNIX OR MINGW OR CYGWIN)
-    SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11 -pthread -O3 -Wextra -Wall -Wno-ignored-attributes -Wno-unknown-pragmas")
+    SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11 -pthread -O2 -g -Wextra -Wall -Wno-ignored-attributes -Wno-unknown-pragmas")
```
Comment (Laurae2) [inline CMakeLists.txt]: Is it really needed?

File: examples/python-guide/sklearn_example.py [hunk 2]
```diff
@@ -17,7 +17,8 @@
 
 print('Start training...')
 # train
-gbm = lgb.LGBMRegressor(objective='regression',
+gbm = lgb.LGBMRegressor(boosting_type='rgf',
```
Comment (Laurae2) [inline examples/python-guide/sklearn_example.py]: I think it's better to create a new example

File: examples/regression/train.conf [hunk 3]
```diff
@@ -2,7 +2,7 @@
 task = train
 
 # boosting type, support gbdt for now, alias: boosting, boost
-boosting_type = gbdt
+boosting_type = rgf
```
Comment (Laurae2) [inline examples/regression/train.conf]: I think it's better to create a new example

**Comment by fukatani** [comment]
Thank you!
This PR still be low completeness.
The fitting score improvement has not been confirmed and rgf crash with large data.

So it can be too early to start a code review. (Or once should I close this PR?)

**Comment by fukatani** [comment]
Sorry, I could not confirm performance improvement.
My implementation can be wrong.
I want to continue this work, but once close this PR.

**Pull request closed without merging**